### PR TITLE
DRAFT: filtering bad course data out of the update contentmetadata from discovery task

### PR DESCRIPTION
**Pull request by alex-sheehan-edx**
I'm wondering if we should read the query and check if the course is valid based on the queries specific `availability` and `status` or if we should just have a base "there needs to be a minimum requirement for the content to be added to the catalog service" like I have now.

## Ticket Link

Link to the associated ticket
https://2u-internal.atlassian.net/browse/ENT-6041

## Post-review

* [ ] Squash commits into discrete sets of changes
* [ ] Ensure that once the changes have been [deployed to stage](https://gocd.tools.edx.org/go/pipeline/activity/stage-enterprise_catalog), prod is manually deployed

### CODE REVIEW

File: enterprise_catalog/apps/catalog/models.py [hunk 1]
```diff
@@ -900,6 +900,33 @@ def update_contentmetadata_from_discovery(catalog_query):
...
+                        course_run.get('status') == "published"):
+                    course_valid = True
+
+            if not course_valid:
+                metadata.pop(index)
+                for key in content_item.get('course_run_keys'):
+                    invalid_course_runs.add(key)
+
+    if invalid_course_runs:
+        num_of_things_to_remove = len(invalid_course_runs)
```
Comment (alex-sheehan-edx) [inline enterprise_catalog/apps/catalog/models.py]: I wonder if I can be smarter here... maybe sort metadata...

File: enterprise_catalog/apps/catalog/models.py [hunk 2]
```diff
@@ -900,6 +900,33 @@ def update_contentmetadata_from_discovery(catalog_query):
     # metadata will be an empty dict if unavailable from cache or API.
     metadata = CatalogQueryMetadata(catalog_query).metadata
 
+    invalid_course_runs = set()
+    for index, content_item in enumerate(metadata):
+        if get_content_type(content_item) == 'course':
+            course_runs = content_item.get('course_runs')
+            course_valid = False
+            for course_run in course_runs:
+                if ((course_run.get('availability') in ["Current", "Starting Soon", "Upcoming"]) and
```
Comment (johnnagro) [inline enterprise_catalog/apps/catalog/models.py]: lets make a validity function to call here

Comment (alex-sheehan-edx) [reply enterprise_catalog/apps/catalog/models.py]: I made some adjustments based on the team's thoughts. It's been pulled out to a method tied to the query

File: enterprise_catalog/apps/catalog/models.py [hunk 3]
```diff
@@ -116,6 +116,38 @@ def get_by_uuid(cls, uuid):
...
             return None
 
+    def is_course_run_content_valid(self, course_run_content):
+        """
+        Check if a piece of course run content (JSON blob) adheres to a query's filter for availability and status
+        """
+        if self.content_filter.get('availability'):
+            if not course_run_content.get('availability') or \
+            course_run_content.get('availability') not in self.content_filter.get('availability'):
+                return False
```
Comment (iloveagent57) [inline enterprise_catalog/apps/catalog/models.py:126]: nit: might be easier to read as
```python
required_availability = self.content_filter.get('availability')
actual_availability = course_run_content.get('availability')
if required_availability and (actual_availability != required_availability):
    return False
```

**Review by iloveagent57 — COMMENTED** [review]
I wonder if we could make this more organized in the following way:
1. Create a new file called `content_validators.py` or something that contains all our filter functions.  They'd generally take a single piece of JSON metadata as input and boolean output.
2. They should be as small as possible.  Even your one-line check in here insisting that there are course runs in a course should be a function.  Like `course_has_runs(content)`.
3. Declare a dict mapping content types to listS of validators to run the content through.  Something like `VALIDATORS_BY_CONTENT_TYPE`.
4. Then in `update_contentmetadata_from_discovery()` you can do something like: 
```python
validated_metadata = []
for record in metadata:
    validators = VALIDATORS_BY_CONTENT_TYPE[get_content_type(record)]
    if all([validator(record) for validator in validators]):
        validated_metadata.append(record)
```

This will have the effect of:
* making it obvious where new validation functions should live
* making it easy, at a glance, to tell how metadata straight from discovery might be filtered out before getting added to the enterprise-catalog models.
* make it easy to plug in new validators - just write your function and add it to that mapping, rather than trying to find the right place in the pipeline to insert any new business logic.

If we do something like this, we should do a search for other instances of content validation that should be explicitly represented in the same way.

**Comment by johnnagro** [comment]
I really like the approach @iloveagent57 outlined in https://github.com/openedx/enterprise-catalog/pull/492#pullrequestreview-1092426801

**Comment by alex-sheehan-edx** [comment]
This PR was made as part of a discovery investigation into filtering bad content. Now that we've reached a satisfying stopping point, I'm going to close this PR and let whoever takes the follow up ticket go off the work/comments here.

**Pull request closed without merging**